include/onnxruntime/core/providers/coreml/coreml_provider_factory.h

```diff
@@ -61,6 +61,22 @@ static const char* const kCoremlProviderOption_SpecializationStrategy = "Special
 static const char* const kCoremlProviderOption_ProfileComputePlan = "ProfileComputePlan";
 // please refer to https://developer.apple.com/documentation/coreml/mlmodelconfiguration/allowlowprecisionaccumulationongpu
 static const char* const kCoremlProviderOption_AllowLowPrecisionAccumulationOnGPU = "AllowLowPrecisionAccumulationOnGPU";
+// Specify the directory to cache any CoreML models created from the ONNX model in.
+// CoreML EP will convert onnx subgraph to CoreML model and save to disk.
+// If this path is not specified, the model will be saved to a temp directory and deleted after the session is closed.
+// otherwise, the model will be saved to the specified path and User should manage to delete the model.
+
+// we would not detect if the cached model match the onnx subgraph, so User should carefully manage the cache for a new model.
+// The cache key is generated by
+//    1. user provided key in metadata_props if exist (prefered)
+//    2. Hash of the model url the inference session was created with
+//    3. Hash of the input/output names of the model
+
+// EP wounln't track the model change or responsible for the cache management.
+static const char* const kCoremlProviderOption_ModelCacheDirectory = "ModelCacheDirectory";
+
+// User provided cache-key in metadata_props.
+static const char* const kCOREML_CACHE_KEY = "CACHE_KEY";
```

Comment:
This is going in model.metadata_props so a very generic 'CACHE_KEY' doesn't convey to someone viewing the model what the value means. If we add more generic caching across ORT as a whole storing it under 'CACHE_KEY' would make more sense.

```suggestion
static const char* const kCOREML_CACHE_KEY = "COREML_CACHE_KEY";
```